File: js/lowest-common-ancestor-of-a-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @param {TreeNode} p
 * @param {TreeNode} q
 * @return {TreeNode}
 */

const mp = {},
      mq = {}

const contains = (r) => mp[r.val] && mq[r.val]

const dfs = (r, p, q) => {
    if (!r) return;
    mp[r.val] = r.val == p
    mq[r.val] = r.val == q
    
    if (r.left) {
        dfs(r.left, p, q)
        mp[r.val] = mp[r.val] || mp[r.left.val]
        mq[r.val] = mq[r.val] || mq[r.left.val]
    }
    
    if (r.right) {
        dfs(r.right, p, q)
        mp[r.val] = mp[r.val] || mp[r.right.val]
        mq[r.val] = mq[r.val] || mq[r.right.val]
    }
}

var lowestCommonAncestor = function(root, p, q) {
    dfs(root, p.val, q.val)
    let res = root
    for (;;) {
        if (res.left && contains(res.left)) res = res.left
        else if (res.right && contains(res.right)) res = res.right
        else break
    }
    return res
};

File: go/lowest-common-ancestor-of-a-binary-tree.go
/**
 * Definition for TreeNode.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
var mp, mq map[int]bool

func contains(r *TreeNode) bool {
	return mp[r.Val] && mq[r.Val]
}

func dfs(r *TreeNode, p, q int) {
	if r == nil {
		return
	}

	mp[r.Val] = r.Val == p
	mq[r.Val] = r.Val == q

	if r.Left != nil {
		dfs(r.Left, p, q)
		mp[r.Val] = mp[r.Val] || mp[r.Left.Val]
		mq[r.Val] = mq[r.Val] || mq[r.Left.Val]
	}

	if r.Right != nil {
		dfs(r.Right, p, q)
		mp[r.Val] = mp[r.Val] || mp[r.Right.Val]
		mq[r.Val] = mq[r.Val] || mq[r.Right.Val]
	}
}

func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	mp = map[int]bool{}
	mq = map[int]bool{}

	dfs(root, p.Val, q.Val)
	res := root
	for {
		if res.Left != nil && contains(res.Left) {
			res = res.Left
		} else if res.Right != nil && contains(res.Right) {
			res = res.Right
		} else {
			break
		}
	}
	return res
}